src/query/storage/completed_tags.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"bytes"
+	"errors"
+	"sort"
+)
+
+// MergeWith merges two CompleteTagsResults together, handling deduplication.
+// Once all merges are completed, Finalize() should be called to give sorted results.
+func (r *CompleteTagsResult) MergeWith(other *CompleteTagsResult) error {
+	nameOnly := r.CompleteNameOnly
+	if nameOnly != other.CompleteNameOnly {
+		return errors.New("cannot merge tag results with name only tag results")
+	}
+
+	if !r.merged {
+		r.merged = true
+		r.seenMap = make(map[string]int, len(r.CompletedTags))
+		for i, tag := range r.CompletedTags {
+			r.seenMap[string(tag.Name)] = i
```

Comment:
interestingly this is optimized by Go: https://github.com/golang/go/issues/3512

Reply:
Nice, as far as I can tell this implies the general pattern is to use `map[string(name)]` rather than using
```
n := string(name)
map[n]
```
right?

Reply:
yup